### ISSUE TITLE
Typed version of vnode-helper

### ISSUE THREAD

**Issue by HerringtonDarkholme** [ktsn/vue-vnode-helper#9]
Thanks for building this library! I find vnode-helper might be the fastest way we can achieve type safety in render function! if we can get vnode-helper typed. 

So let me provide some background. Vue itself has many libraries for its type checking. But vue's template has very few options to type checking. tsx is actually not an option because TypeScript only supports react compatible format. (more info [here](https://github.com/HerringtonDarkholme/av-ts/issues/13#issuecomment-260233573))

However, we might change vnode-helper's type signature to enable type checking in plain ts.

For example, given a component definition like below.

```ts
class VueComp extends Vue {
  // a special field for props type declaration
  props: {
     name: string
  }
  // a special field for event type declaration
  $emit: {
     event: string
  }
}
```

We can generate a helper method having the signature:

```ts
function tag(name: 'VueComp', options: {
  props?: {
    name?: string
  },
  on?: {
    event?(payload: string): void
  }
}): vnode {
  // ....
}
```

In use.

```ts
tag('VueComp', {props: {name: 'ktsn'}}) // optional event
tag('VueComp', {prop: {name: 'hchan'}}) // error, typo in prop
tag('VueComp', {on: {event(n: number) {}}}) // error, type mismatch
```

I'm keenly looking forward to your "Vue" point :)

**Comment by ktsn** [comment]
Hi, thanks for your idea :)
I like the idea of using string literal type to infer a component `props` and `on`.
I actually thought a similar thing about type safety of the render function but it uses component's constructor only. [Example code](https://www.typescriptlang.org/play/#src=interface%20VueStatic%3CP%2C%20E%3E%20%7B%0D%0A%20%20new%20%28%29%3A%20Vue%3CP%2C%20E%3E%0D%0A%7D%0D%0A%0D%0Aclass%20Vue%3CP%2C%20E%3E%20%7B%0D%0A%20%20constructor%28%29%20%7B%7D%0D%0A%20%20_propsBrand%3A%20P%0D%0A%20%20_eventsBrand%3A%20E%0D%0A%7D%0D%0A%0D%0Atype%20ClassType%3CP%2C%20E%2C%20T%20extends%20Vue%3CP%2C%20E%3E%2C%20C%20extends%20VueStatic%3CP%2C%20E%3E%3E%20%3D%0D%0A%20%20C%20%26%0D%0A%20%20%28new%20%28%29%20%3D%3E%20T%29%20%26%0D%0A%20%20%28new%20%28%29%20%3D%3E%20%7B%20_propsBrand%3A%20P%2C%20_eventsBrand%3A%20E%20%7D%29%0D%0A%0D%0Ainterface%20CreateElement%20%7B%0D%0A%20%20%3CP%2C%20E%2C%20T%20extends%20Vue%3CP%2C%20E%3E%2C%20C%20extends%20VueStatic%3CP%2C%20E%3E%3E%28%0D%0A%20%20%20%20type%3A%20ClassType%3CP%2C%20E%2C%20T%2C%20C%3E%2C%0D%0A%20%20%20%20data%3F%3A%20%7B%20props%3F%3A%20P%2C%20on%3F%3A%20E%20%7D%2C%0D%0A%20%20%20%20children%3F%3A%20any%5B%5D%0D%0A%20%20%29%3A%20any%0D%0A%7D%0D%0A%0D%0A%2F%2F%20------------------------------------------------------------------%0D%0A%0D%0Ainterface%20FooProps%20%7B%0D%0A%20%20foo%3A%20string%0D%0A%20%20bar%3A%20number%0D%0A%7D%0D%0A%0D%0Ainterface%20FooEvents%20%7B%0D%0A%20%20test%3A%20%28event%3A%20Event%29%20%3D%3E%20void%0D%0A%7D%0D%0A%0D%0Aclass%20FooComponent%20extends%20Vue%3CFooProps%2C%20FooEvents%3E%20implements%20FooProps%20%7B%0D%0A%20%20foo%3A%20string%0D%0A%20%20bar%3A%20number%0D%0A%0D%0A%20%20hoge%28%29%20%7B%7D%0D%0A%7D%0D%0A%0D%0Aexport%20default%20%7B%0D%0A%20%20render%20%28h%3A%20CreateElement%29%20%7B%0D%0A%20%20%20%20return%20h%28FooComponent%2C%20%7B%0D%0A%20%20%20%20%20%20props%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20foo%3A%20'foo'%2C%0D%0A%20%20%20%20%20%20%20%20bar%3A%201%0D%0A%20%20%20%20%20%20%7D%2C%0D%0A%20%20%20%20%20%20on%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20test%28event%3A%20Event%29%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20console.log%28event%29%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%29%0D%0A%20%20%7D%0D%0A%7D%0D%0A)

But I'm not sure how the users register the component types to vnode-helper function. I don't want the users to write the types by themselves. Do you have any ideas about this?

I'm thinking one alternative to expose the [`create` helper](https://github.com/ktsn/vue-vnode-helper/blob/master/src/helper.js#L5) and let users create their helpers that correspond with each component. I think we can generate typed helpers automatically in this approach.

``` js
interface Props {
  name: string
}

interface Events {
  event (payload: string): void
}

class MyComp extends Vue<Props, Events> {
}

// myComp should be inferred Props and Events types for their expected data
export const myComp = create(MyComp)
```

**Comment by HerringtonDarkholme** [comment]
> It uses component's constructor only

IMHO this is even better, using string as component tag is mainly for vue native template. For hand written render function string component tag is not needed.

**Comment by ktsn** [comment]
Indeed, I'll refactoring this library with TypeScript and implement this feature.

**Comment by HerringtonDarkholme** [comment]
Would you mind delaying the refactor? 

TypeScript recently has introduced new features like indexed access type and mapped types. They are very promising and might remove the generic parameter needed to specify props.

**Comment by ktsn** [comment]
Yeah, the new features are very nice 😄 
I wait for next ts release.

**Comment by HerringtonDarkholme** [comment]
I have tried to translate some code into `createElement` and `vnode` helper style.
https://gist.github.com/HerringtonDarkholme/b3fccccf1abe7cde27686f613f53d55a

I have to say the ending result is not satisfactory. (At least I don't want to bother write so many parenthesis/brackets for type safety 😞 ).

The main problem is readability. In vue template, every tag is highlighted, and properties does not need any curly brackets to delimit them. However, in plain TS, function call is not well highlighted. Lots of function call and object literal are more like visual cruft than helpful guidance, especially when one forgets to type rightly to balance a brackets/parenthesis pair. It just reminds me of Lisp jokes.

If it were python or haskell, alas coffee/pure script, indentation might be ergonomic enough for code template.

Another problem is tooling. In vue template we can use emmet or so, while no snippets for typescript. But users can simply add custom snippets. So it is not hard to solve.

**Comment by ktsn** [comment]
I'm sorry for the late reply!

Hmm, it is less readable exactly. I actually thought the similar feeling when I try this library in my (small) project.
I think what we can do for this problem is split the vnode into smaller components/functions/variables as React's best practice. I feel there is difference about appropriate granularity between template (Vue) and render (React JSX) when I write them. So maybe we can reduce the pain by splitting them. (I'm not sure because I'm not use this in production yet 😅 )

Anyway, I'll start the refactoring from now since the new version of TS is released.

**Comment by HerringtonDarkholme** [comment]
I have recently worked on an alternative syntax for typed template code.  https://github.com/HerringtonDarkholme/vivio

It turns out to be an exploration in how far we can express in TypeScript compiler. The ending result is rather astounding, we can fully express Vue's template logic in TypeScript and check every component properties and event, with similar API surface. Furthermore we can encode the HTML structure at type level, which ensure users return a well-formed html fragment. Say, if user forget to add a closing tag function call, it will become a compiler error. It's my best effort to achieve something similar with Angular's [language](https://github.com/angular/angular/tree/master/modules/%40angular/language-service) [service](https://github.com/angular/angular/issues/7482). I think it requires man power way beyond OSS to provide a true, robust language service for Vue, sadly.

It's too experimental in implementation and heretic in design. Vnode-helper's API is better in common sense. But I think it might be a reference for implementing type safe template. The type definition is very daunting, though 😞. https://github.com/HerringtonDarkholme/vivio/blob/master/src/component.ts

 I hope vnode-helper can give us a more canonical API and type safety at the same time 😄

**Comment by ktsn** [comment]
Oh, it's very interesting!
I'll learn from your idea and code 😄  Thanks for sharing!

**Comment by ktsn** [comment]
@HerringtonDarkholme 
Hi! I'd released [v0.3.0](https://github.com/ktsn/vue-vnode-helper/releases/tag/v0.3.0) with TS declaration files. And I declare [vnode-helper's type as an interface](https://github.com/ktsn/vue-vnode-helper/blob/master/src/declarations.ts#L18-L28), so we can augment the vnode-helper type.

I ended up separating the code that recognize components' `props` and `on` types because there might be various approach for that. So, I've create [an experimental project using vue-vnode-helper](https://github.com/ktsn/vue-comm) that extends vue-class-component API. It can successfully check `props` and `on` for simple case.

Thanks for suggesting typed version of vnode helper 😄
